yamux/src/connection/stream/flow_control.rs

```diff
@@ -0,0 +1,305 @@
+        //
+        // In case the remote sender has exhausted half or more of its credit in less than 2
+        // round-trips, try to double `max_receive_window_size`.
+        //
+        // For simplicity `max_receive_window_size` is never decreased.
+        //
+        // This implementation is heavily influenced by QUIC. See document below for rational on the
+        // above strategy.
+        //
+        // https://docs.google.com/document/d/1F2YfdDXKpy20WVKJueEf4abn_LVZHhMUMS5gX6Pgjl4/edit?usp=sharing
+        if self
+            .rtt
+            .get()
+            .map(|rtt| self.last_window_update.elapsed() < rtt * 2)
+            .unwrap_or(false)
+        {
+            let mut accumulated_max_stream_windows = self.accumulated_max_stream_windows.lock();
+
+            // Ideally one can just double it:
+            let mut new_max = self.max_receive_window_size.saturating_mul(2);
+
+            // But one has to consider the configured connection limit:
+            new_max = {
+                let connection_limit: usize = self.max_receive_window_size as usize +
+                    // the overall configured conneciton limit
+                    (self.config.max_connection_receive_window.unwrap_or(usize::MAX)
+                    // minus the minimum amount of window guaranteed to each stream
+                    - self.config.max_num_streams * DEFAULT_CREDIT as usize
+                    // minus the amount of bytes beyond the minimum amount (`DEFAULT_CREDIT`)
+                    // already allocated by this and other streams on the connection.
+                    - *accumulated_max_stream_windows);
+
+                cmp::min(new_max, connection_limit.try_into().unwrap_or(u32::MAX))
+            };
+
+            // Account for the additional credit on the accumulated connection counter.
+            *accumulated_max_stream_windows += (new_max - self.max_receive_window_size) as usize;
+            drop(accumulated_max_stream_windows);
+
+            log::debug!(
+                "old window_max: {} mb, new window_max: {} mb",
+                self.max_receive_window_size as f64 / 1024.0 / 1024.0,
+                new_max as f64 / 1024.0 / 1024.0
+            );
+
+            self.max_receive_window_size = new_max;
+
+            // Recalculate `next_window_update` with the new `max_receive_window_size`.
+            let bytes_received = self.max_receive_window_size - self.current_receive_window_size;
+            next_window_update =
+                bytes_received.saturating_sub(buffer_len.try_into().unwrap_or(u32::MAX));
+        }
+
+        self.last_window_update = Instant::now();
+        self.current_receive_window_size += next_window_update;
+
+        self.assert_invariants(buffer_len);
+
+        return Some(next_window_update);
+    }
+
+    fn assert_invariants(&self, buffer_len: usize) {
+        if !cfg!(debug_assertions) {
+            return;
+        }
```

Comment:
I find this odd, why not just use `debug_assert` further down? If this is an optimisation, it at least needs a comment.

Reply:
Because `rtt.get()` and `self.accumulated_max_stream_windows.lock()` both take a lock. I would like to not do so in non-debug.